src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStep.java

```diff
@@ -0,0 +1,512 @@
+  protected final GcsPdao gcsPdao;
+  protected final ObjectMapper objectMapper;
+  protected final Dataset dataset;
+  protected final int maxFailedFileLoads;
+  protected final int maxBadLoadFileLineErrorsReported;
+  protected final FireStoreDao fileDao;
+  protected final FileService fileService;
+  protected final ExecutorService executor;
+  protected final int maxPerformanceThreadQueueSize;
+
+  // What will ultimately be returned to the end user via flight results
+  private final BulkLoadResultModel resultModel = new BulkLoadResultModel();
+  private final List<BulkLoadFileResultModel> loadedFiles = new ArrayList<>();
+
+  public IngestBulkGcpStep(
+      String loadTag,
+      UUID profileId,
+      AuthenticatedUserRequest userReq,
+      GcsPdao gcsPdao,
+      ObjectMapper objectMapper,
+      Dataset dataset,
+      int maxFailedFileLoads,
+      int maxBadLoadFileLineErrorsReported,
+      FireStoreDao fileDao,
+      FileService fileService,
+      ExecutorService executor,
+      int maxPerformanceThreadQueueSize) {
+    this.loadTag = loadTag;
+    this.profileId = profileId;
+    this.userReq = userReq;
+    this.gcsPdao = gcsPdao;
+    this.objectMapper = objectMapper;
+    this.dataset = dataset;
+    this.maxFailedFileLoads = maxFailedFileLoads;
+    this.maxBadLoadFileLineErrorsReported = maxBadLoadFileLineErrorsReported;
+    this.fileDao = fileDao;
+    this.fileService = fileService;
+    this.executor = executor;
+    this.maxPerformanceThreadQueueSize = maxPerformanceThreadQueueSize;
+  }
+
+  /**
+   * Return a stream of file models representing the files to ingest. The construction of this
+   * stream varies based on the type of ingestion flight
+   */
+  protected abstract Stream<BulkLoadFileModel> getModelsStream(FlightContext context);
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+
+    initializeResultsModel(context.getFlightId(), loadTag);
+
+    FlightMap workingMap = context.getWorkingMap();
+
+    GoogleBucketResource bucketResource =
+        FlightUtils.getContextValue(context, FileMapKeys.BUCKET_INFO, GoogleBucketResource.class);
+    List<FileModel> ingestedFiles;
+    try (var loadModels = getModelsStream(context)) {
+      ingestedFiles = performIngest(workingMap, loadModels.toList(), bucketResource);
+    } catch (InvalidUserProjectException ex) {
+      // We retry this exception because often when we've seen this error it has been transient
+      // and untruthful -- i.e. the user project specified exists and has a legal id.
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    } catch (GoogleInternalServerErrorException ex) {
+      // Google's error message suggests retrying the operation
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    } catch (FileSystemAbortTransactionException rex) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, rex);
```

Comment:
In attempting to compare this step to the existing code for ingest, I see there is a lot of logic in the[ old implementation](https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/filedata/flight/ingest/IngestDriverStep.java#L126) that handles the case on retry where an ingest is in progress or has been orphaned. 

You mentioned today that this flight should be idempotent. How would it handle the case that a file is in progress of being ingested when we retry the ingest? Could we end up with duplicate files?

Reply:
That's a great question.  We shouldn't because even if we do regenerate file ids - as long as the paths match, we end up grabbing the id from the path from a previous ingest (that's one of the differences with the bulk upload method you commented about above). That's where the idconflicts map comes into play